tool/mustachio/codegen_aot_compiler.dart

```diff
@@ -0,0 +1,473 @@
+        continue;
+      }
+
+      var type = getter.returnType;
+      var contextChain = '${context.name}.$primaryName';
+      var remainingNames = [...key.skip(1)];
+      for (var secondaryKey in remainingNames) {
+        getter = (type as InterfaceType)
+            .lookUpGetter2(secondaryKey, type.element.library);
+        if (getter == null) {
+          throw MustachioResolutionError(node.keySpan.message(
+              "Failed to resolve '$secondaryKey' on ${context.type} while "
+              'resolving $remainingNames as a property chain on any types in '
+              'the context chain: $contextChain, after first resolving '
+              "'$primaryName' to a property on $type"));
+        }
+        type = getter.returnType;
+        contextChain = '$contextChain.$secondaryKey';
+      }
+      return _VariableLookup(type, contextChain);
+    }
+
+    var contextTypes = [
+      for (var c in _contextStack) c.type,
+    ];
+    throw MustachioResolutionError(node.keySpan
+        .message("Failed to resolve '$key' as a property on any types in the "
+            'context chain: $contextTypes'));
+  }
+
+  /// Writes [content] to the generated render functions as text, properly
+  /// handling newlines, quotes, and other special characters.
+  void _writeText(String content) {
+    if (content.isEmpty) return;
+    if (content == '\n') {
+      // Blank lines happen a lot; just indicate them as such.
+      writeln('buffer.writeln();');
+    } else {
+      content = content.replaceAll("'", "\\'").replaceAll(r'$', r'\\$');
+      if (RegExp('^[ \\n]+\$').hasMatch(content)) {
+        write("buffer.write('");
+        write(content.replaceAll('\n', '\\n'));
+        writeln("');");
+      } else {
+        write("buffer.write('''");
+        write(content);
+        writeln("''');");
+      }
+    }
+  }
+
+  /// Writes a call to [variableLookup] to the renderer.
+  ///
+  /// The result is HTML-escaped if [escape] is true.
+  void _writeGetter(_VariableLookup variableLookup, {bool escape = true}) {
+    var variableAccess = variableLookup.name;
+    if (escape) {
+      writeln('buffer.write(htmlEscape.convert('
+          '$variableAccess.toString()));');
+    } else {
+      writeln('buffer.write($variableAccess.toString());');
+    }
+  }
+}
+
+/// Various static build data to be used for each renderer, including specified
+/// renderers and template renderers.
+@immutable
+class _BuildData {
+  final BuildStep _buildStep;
+
+  final TypeProvider _typeProvider;
+
+  final TypeSystem _typeSystem;
+
+  final TemplateFormat _format;
+
+  _BuildData(
+      this._buildStep, this._typeProvider, this._typeSystem, this._format);
+}
+
+/// Represents a variable lookup via property access chain [name] which returns
+/// an object of type [type].
+@immutable
+class _VariableLookup {
+  final InterfaceType type;
+
+  final String name;
+
+  _VariableLookup(this.type, this.name);
+}
+
+extension<T> on List<T> {
```

Comment:
Had no idea you could do `extension<T>` but I guess it makes sense!

Reply:
yes very neat!